File: tests/benchmarks/node/parse.js
function nowMs() {
  return Number(process.hrtime.bigint() / 1000000n);
}

function benchJsonParse(n) {
  const payload = {};
  for (let i = 0; i < n; i++) payload["k" + i] = i;
  const s = JSON.stringify(payload);
  const t0 = nowMs();
  const obj = JSON.parse(s);
  const t1 = nowMs();
  return { case: "json-parse", scale: n, result: Object.keys(obj).length, duration_ms: t1 - t0 };
}

function genJsCode(n) {
  let out = "";
  for (let i = 0; i < n; i++) out += `var a${i}=${i};\n`;
  return out;
}

function benchCompileFunction(n) {
  const code = genJsCode(n);
  const t0 = nowMs();
  // new Function compiles the code; we do not execute it.
  const fn = new Function(code);
  const t1 = nowMs();
  return { case: "compile-fn", scale: n, result: n, duration_ms: t1 - t0 };
}

const args = (() => {
  let scale = 5000;
  for (let i = 2; i < process.argv.length; i++) {
    if (process.argv[i] === "--scale") scale = parseInt(process.argv[++i], 10);
  }
  return { scale };
})();

const n = args.scale;
const out = [benchJsonParse(n), benchCompileFunction(n)];
for (const item of out) console.log(JSON.stringify(item));

File: tests/benchmarks/python/parse.py
import argparse
import time
import json

def gen_python_code(n: int) -> str:
    # generate sequential assignments: a0=0; a1=1; ...
    lines = [f"a{i}={i}" for i in range(n)]
    return "\n".join(lines)

def bench_json_parse(n: int):
    payload = {f"k{i}": i for i in range(n)}
    s = json.dumps(payload, ensure_ascii=False)
    t0 = time.perf_counter()
    obj = json.loads(s)
    t1 = time.perf_counter()
    return {"case": "json-parse", "scale": n, "result": len(obj), "duration_ms": (t1 - t0) * 1000.0}

def bench_ast_parse(n: int):
    import ast
    code = gen_python_code(n)
    t0 = time.perf_counter()
    tree = ast.parse(code)
    t1 = time.perf_counter()
    return {"case": "ast-parse", "scale": n, "result": len(getattr(tree, 'body', [])), "duration_ms": (t1 - t0) * 1000.0}

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--scale", type=int, default=5000)
    args = ap.parse_args()
    out = [bench_json_parse(args.scale), bench_ast_parse(args.scale)]
    for item in out:
        print(json.dumps(item, ensure_ascii=False))

if __name__ == "__main__":
    main()

